### DISCUSSION TITLE
Error: Cannot find external module "buffer" - what to do?

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#9760] [Q&A]
In https://github.com/hindley-milner-systems/dapp-ertp-airdrop/issues/72#issuecomment-2245669503 , we see

```
Error {
    message: 'Cannot find external module "buffer" in package file:///Users/tgreco/merkle-airdrop-resources/dapp-ertp-airdrop/node_modules/merkletreejs/',
}
```

What's going on? What do we do?

cc @tgrecojs @Jovonni

**Comment by dckc** [comment]
Buffer is a Node.js host API...

> The conventional globals defined by browser or Node.js hosts are not available by default... `Buffer` ...
 -- [Hardening JavaScript: Limiting Globals with Compartments](https://docs.agoric.com/guides/js-programming/hardened-js.html#hardening-javascript-limiting-globals-with-compartments)

I think npm has a number of libraries that provide the buffer API in a browser context. Some of them are probably suitable for use in contracts. @Jovonni I wonder if you have a favorite?

@kriskowal I wonder if an `@endo/buffer` API would be cost-effective.

**Comment by tgrecojs** [comment]
With a bit of digging, i've traced the string `Cannot find external module <package name>` back to `@endo/compartment-mapper`. Here's a link to [all files that mention it](https://github.com/search?q=repo%3Aendojs%2Fendo+%22cannot+find+external+module%22&type=code).

**Comment by dckc** [comment]
Yes, the compartment mapper is **correctly** complaining that the `merkletreejs` package refers to a `buffer` package:

```js
import { Buffer } from 'buffer'
```
 -- https://github.com/merkletreejs/merkletreejs/blob/13b886e0f02c5b90369e5bf0f539aa3fdc769f8b/src/MerkleTree.ts#L1

but it does not declare a dependency on any such package:

```
  "dependencies": {
    "bignumber.js": "^9.0.1",
    "buffer-reverse": "^1.0.1",
    "crypto-js": "^4.2.0",
    "treeify": "^1.1.0",
    "web3-utils": "^1.3.4"
  },
```
 -- https://github.com/merkletreejs/merkletreejs/blob/13b886e0f02c5b90369e5bf0f539aa3fdc769f8b/package.json#L87-L93


The `merkletreejs` package assumes the Node.js [buffer](https://nodejs.org/api/buffer.html) host API. That Node.js host API is not provided by the Agoric smart contract runtime.
